### PR TITLE
KBV-771: Updating the stub to more easily test multiple addresses

### PR DESCRIPTION
**Pull request by gds-jbentham**
### What changed

Updating the stub so valid from and valid until can be set and a second address can be added under new user

### Why did it change

To more easily test edge cases regarding multiple address at an individual CRI level

### Issue tracking

- [KBV-771](https://govukverify.atlassian.net/browse/KBV-771)

### CODE REVIEW

**Review by smsgds — DISMISSED** [review]

File: di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java [hunk 1]
```diff
@@ -1,12 +1,14 @@
 package uk.gov.di.ipv.stub.core.config.uatuser;
 
 import spark.QueryParamsMap;
+import spark.utils.StringUtils;
 
-import java.time.Instant;
-import java.time.LocalDate;
-import java.time.ZoneId;
+import java.time.*;
```
Comment (tal-nagra) [inline di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java]: avoid the use of the .* imports and reinstate the previous import statements.
Intellij IDEA has a setting that needs to be configured to prevent the `*` automatically being introduced

File: di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java [hunk 2]
```diff
@@ -51,7 +53,8 @@ public Identity map(Map<String, String> map, int rowNumber) {
                         map.get("district"),
                         map.get("posttown"),
                         map.get("postcode"),
-                        true);
+                        LocalDate.of(2021, 1, 1),
```
Comment (tal-nagra) [inline di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java]: It might be worth declaring: `LocalDate addressValidFrom = LocalDate.of(2021, 1, 1);` and referencing it here to make it more obvious what the date means

File: di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/Identity.java [hunk 3]
```diff
@@ -1,10 +1,12 @@
 package uk.gov.di.ipv.stub.core.config.uatuser;
 
+import java.util.List;
+
 public record Identity(
         int rowNumber,
         String accountNumber,
         String ctdbDatabase,
-        UKAddress UKAddress,
+        List<UKAddress> UKAddress,
```
Comment (tal-nagra) [inline di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/Identity.java]: rename to `addresses`. At present only UK addresses are supported therefore the UK prefix on the variable name doesn't really add any value

File: di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java [hunk 4]
```diff
@@ -78,6 +81,41 @@ public DisplayIdentity mapToDisplayable(Identity identity) {
 
     public SharedClaims mapToSharedClaim(Identity identity, boolean agedDOB) {
         FindDateOfBirth dateOfBirth = identity.findDateOfBirth();
+
+        List<CanonicalAddress> canonicalAddresses = new ArrayList<>();
+        for (UKAddress address : identity.UKAddress()) {
+            LocalDate validFrom = address.validFrom();
```
Comment (tal-nagra) [inline di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java]: There is no need to declare the validFrom and validUntil variables.

File: di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java [hunk 5]
```diff
@@ -78,6 +81,41 @@ public DisplayIdentity mapToDisplayable(Identity identity) {
 
     public SharedClaims mapToSharedClaim(Identity identity, boolean agedDOB) {
         FindDateOfBirth dateOfBirth = identity.findDateOfBirth();
+
+        List<CanonicalAddress> canonicalAddresses = new ArrayList<>();
+        for (UKAddress address : identity.UKAddress()) {
```
Comment (tal-nagra) [inline di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java]: replace the for loop with identity.addresses().stream().map().collect() etc

File: di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java [hunk 6]
```diff
@@ -78,6 +81,41 @@ public DisplayIdentity mapToDisplayable(Identity identity) {
...
+            CanonicalAddress canonicalAddress =
+                    new CanonicalAddress(
+                            address.buildingNumber(),
+                            address.buildingName(),
+                            address.street(),
+                            address.townCity(),
+                            address.postCode(),
+                            // default / arbitrary value assigned for now as
+                            // the validFrom date is not available in test data
+                            validFrom,
```
Comment (tal-nagra) [inline di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java]: simply reference address.validFrom() here

File: di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java [hunk 7]
```diff
@@ -78,6 +81,41 @@ public DisplayIdentity mapToDisplayable(Identity identity) {
...
+                            validUntil);
```
Comment (tal-nagra) [inline di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java]: change to address.validUntil()

File: di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java [hunk 8]
```diff
@@ -78,6 +81,41 @@ public DisplayIdentity mapToDisplayable(Identity identity) {
...
+            canonicalAddresses.add(canonicalAddress);
+        }
+
+        if (canonicalAddresses.get(0).validFrom() == null) {
```
Comment (tal-nagra) [inline di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java]: Remove this conditional statement code block 
Add the validFrom default date assignment logic above ( when mapping the identity.addresses() to List<CanonicalAddress>

File: di-ipv-core-stub/src/main/resources/templates/edit-user.mustache [hunk 9]
```diff
@@ -184,6 +184,236 @@
 
             </div>
 
+            <div class="govuk-date-input" id="validFrom">
+
+                <div class="govuk-date-input__item">
+                    <div class="govuk-form-group">
+                        <label class="govuk-label govuk-date-input__label" for="validFrom-day">
```
Comment (tal-nagra) [inline di-ipv-core-stub/src/main/resources/templates/edit-user.mustache]: use camelCase or hyphenated identifiers for all id/name attribute values in this file
e.g. validFromDay 
or
valid-from-day

File: di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java [hunk 10]
```diff
@@ -114,6 +142,11 @@ public List<QuestionAndAnswer> mapToQuestionAnswers(
     }
 
     public Identity mapFormToIdentity(Identity identityOnRecord, QueryParamsMap userData) {
```
Comment (tal-nagra) [inline di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java]: rename `userData` to `formData`

File: di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java [hunk 11]
```diff
@@ -114,6 +142,11 @@ public List<QuestionAndAnswer> mapToQuestionAnswers(
...
+        List<UKAddress> addresses = new ArrayList<>();
+
+        LocalDate vf = getLocalDate(userData, "validFrom-year", "validFrom-month", "validFrom-day");
+
+        LocalDate vu = getLocalDate(userData, "validFrom-year", "validFrom-month", "validFrom-day");
 
         UKAddress ukAddress =
```
Comment (tal-nagra) [inline di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java]: rename ukAddress to a more specific name
e.g. `primaryAddress` / `firstAddress` / `currentAddress` etc

File: di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java [hunk 12]
```diff
@@ -114,6 +142,11 @@ public List<QuestionAndAnswer> mapToQuestionAnswers(
     }
 
     public Identity mapFormToIdentity(Identity identityOnRecord, QueryParamsMap userData) {
+        List<UKAddress> addresses = new ArrayList<>();
+
+        LocalDate vf = getLocalDate(userData, "validFrom-year", "validFrom-month", "validFrom-day");
```
Comment (tal-nagra) [inline di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java]: rename the variables to make their meaning/purpose more obvious:
e.g. primaryAddressValidFrom primaryAddressValidUntil

File: di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java [hunk 13]
```diff
@@ -123,23 +156,72 @@ public Identity mapFormToIdentity(Identity identityOnRecord, QueryParamsMap user
...
-                        true);
-
-        int year = Integer.parseInt(userData.value("dateOfBirth-year"));
-        int month = Integer.parseInt(userData.value("dateOfBirth-month"));
-        int dayOfMonth = Integer.parseInt(userData.value("dateOfBirth-day"));
+                        vf,
+                        vu);
+        addresses.add(ukAddress);
+
+        LocalDate secondaryVf =
```
Comment (tal-nagra) [inline di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java]: rename to secondaryAddressValidFrom

File: di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java [hunk 14]
```diff
@@ -123,23 +156,72 @@ public Identity mapFormToIdentity(Identity identityOnRecord, QueryParamsMap user
...
+                getLocalDate(
+                        userData,
+                        "SecondaryUKAddress.validFrom-year",
+                        "SecondaryUKAddress.validFrom-month",
+                        "SecondaryUKAddress.validFrom-day");
+
+        LocalDate secondaryVu =
```
Comment (tal-nagra) [inline di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java]: rename to secondaryAddressValidUntil

File: di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java [hunk 15]
```diff
@@ -123,23 +156,72 @@ public Identity mapFormToIdentity(Identity identityOnRecord, QueryParamsMap user
...
+                getLocalDate(
+                        userData,
+                        "SecondaryUKAddress.validUntil-year",
+                        "SecondaryUKAddress.validUntil-month",
+                        "SecondaryUKAddress.validUntil-day");
+
+        UKAddress secondaryUkAddress =
```
Comment (tal-nagra) [inline di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java]: rename to `secondaryAddress`

File: di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java [hunk 16]
```diff
@@ -123,23 +156,72 @@ public Identity mapFormToIdentity(Identity identityOnRecord, QueryParamsMap user
...
+                new UKAddress(
+                        userData.value("SecondaryUKAddress.buildingNumber"),
+                        userData.value("SecondaryUKAddress.buildingName"),
+                        userData.value("SecondaryUKAddress.street"),
+                        null,
+                        userData.value("SecondaryUKAddress.townCity"),
+                        userData.value("SecondaryUKAddress.postCode"),
+                        secondaryVf,
+                        secondaryVu);
+        if (!Stream.of(
```
Comment (tal-nagra) [inline di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java:184]: move this `if` conditional statement further up and check the form data for the presence of some of the mandatory address fields e.g. street / postcode
Then only parse the dates and instantiate UKAddress if the form data indicates that a secondary address has been entered

File: di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java [hunk 17]
```diff
@@ -123,23 +156,72 @@ public Identity mapFormToIdentity(Identity identityOnRecord, QueryParamsMap user
...
                 identityOnRecord.accountNumber(),
                 identityOnRecord.ctdbDatabase(),
-                ukAddress,
+                addresses,
                 findDateOfBirth,
                 fullName,
                 identityOnRecord.questions());
     }
+
+    private LocalDate getLocalDate(QueryParamsMap userData, String s, String s2, String s3) {
```
Comment (tal-nagra) [inline di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java]: rename 
`s` to `year`
`s2` to `month`
`s3` to `day`
if you think it helps with legibility then add a suffix to the names e.g. `yearFieldName`

`userData` to `formData`

File: di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java [hunk 18]
```diff
@@ -123,23 +156,72 @@ public Identity mapFormToIdentity(Identity identityOnRecord, QueryParamsMap user
...
+        if (!Stream.of(userData.value(s), userData.value(s2), userData.value(s3))
+                .allMatch(StringUtils::isBlank)) {
+            int secondaryVfYear = Integer.parseInt(userData.value(s));
```
Comment (tal-nagra) [inline di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java]: These 3 variables can be removed and the `Integer.parseInt(formData.value("year")` can be inlined when invoking `LocalDate.of()`

**Review by tal-nagra — CHANGES_REQUESTED** [review]

File: di-ipv-core-stub/src/main/resources/templates/edit-user.mustache (hunk 9 above)
Comment (gds-jbentham) [reply di-ipv-core-stub/src/main/resources/templates/edit-user.mustache]: I left it this format to be consistent with what is in for DoB but we can change this now to camel case and just highlight the difference when test pick it up

File: di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java (hunk 16 above)
Comment (gds-jbentham) [reply di-ipv-core-stub/src/main/java/uk/gov/di/ipv/stub/core/config/uatuser/IdentityMapper.java:184]: The reason to check them all is to allow an address to be created with any combination of fields so we can check error scenarios more easily

**Review by tal-nagra — DISMISSED** [review]

**Review by smsgds — APPROVED** [review]
